CODING_STYLE.md

```diff
@@ -0,0 +1,335 @@
+
+#### Use Standard libc Macros When Exiting
+
+- libc provides `EXIT_FAILURE` and `EXIT_SUCCESS`. Use them whenever possible
+  in the child of `fork()`ed process or when exiting from a `main()` function.
+
+#### Use `goto`s
+
+`goto`s are an essential language construct of C and are perfect to perform
+cleanup operations or simplify the logic of functions. However, here are the
+rules to use them:
+- use descriptive `goto` labels.
+  For example, if you know that this label is only used as an error path you
+  should use something like `on_error` instead of `out` as label name.
+- **only** jump downwards unless you are handling `EAGAIN` errors and want to
+  avoid `do-while` constructs.
+- An example of a good usage of `goto` is:
+  ```
+  static int set_config_idmaps(const char *key, const char *value,
+                             struct lxc_conf *lxc_conf, void *data)
+  {
+          unsigned long hostid, nsid, range;
+          char type;
+          int ret;
+          struct lxc_list *idmaplist = NULL;
+          struct id_map *idmap = NULL;
+
+          if (lxc_config_value_empty(value))
+                  return lxc_clear_idmaps(lxc_conf);
+
+          idmaplist = malloc(sizeof(*idmaplist));
+          if (!idmaplist)
+                  goto on_error;
+
+          idmap = malloc(sizeof(*idmap));
+          if (!idmap)
+                  goto on_error;
+          memset(idmap, 0, sizeof(*idmap));
+
+          ret = parse_idmaps(value, &type, &nsid, &hostid, &range);
+          if (ret < 0) {
+                  ERROR("Failed to parse id mappings");
+                  goto on_error;
+          }
+
+          INFO("Read uid map: type %c nsid %lu hostid %lu range %lu", type, nsid, hostid, range);
+          if (type == 'u')
+                  idmap->idtype = ID_TYPE_UID;
+          else if (type == 'g')
+                  idmap->idtype = ID_TYPE_GID;
+          else
+                  goto on_error;
+
+          idmap->hostid = hostid;
+          idmap->nsid = nsid;
+          idmap->range = range;
+          idmaplist->elem = idmap;
+          lxc_list_add_tail(&lxc_conf->id_map, idmaplist);
+
+          if (!lxc_conf->root_nsuid_map && idmap->idtype == ID_TYPE_UID)
+                  if (idmap->nsid == 0)
+                          lxc_conf->root_nsuid_map = idmap;
+
+
+          if (!lxc_conf->root_nsgid_map && idmap->idtype == ID_TYPE_GID)
+                  if (idmap->nsid == 0)
+                          lxc_conf->root_nsgid_map = idmap;
+
+          idmap = NULL;
+
+          return 0;
+
+  on_error:
+          free(idmaplist);
+          free(idmap);
+
+          return -1;
+  }
+  ```
+
+#### Use Booleans instead of integers
+
+- When something can be conceptualized in a binary way use a boolean not
+  an integer.
+
+#### Cleanup Functions Must Handle The Object's `NULL` Type
+
+- If you implement a custom cleanup function to e.g. free a complex type
+  you declared you must ensure that the object's `NULL` type is handled and
```

Comment:
I'm confused by both the wording ("null type") and the example below, as it clearly cannot be called with `NULL` as `p`.

Reply:
Yeah, I should rewrite this